swap/peer.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethersphere/swarm/p2p/protocols"
+)
+
+var SigDoesNotMatch = errors.New("signature does not match")
+var DontOwe = errors.New("no negative balance")
+
+// Peer is a devp2p peer for the Swap protocol
+type Peer struct {
+	*protocols.Peer
+	swap *Swap
+}
+
+func NewPeer(p *protocols.Peer, s *Swap) *Peer {
+	return &Peer{
+		Peer: p,
+		swap: s,
+	}
+}
+
+// handleMsg is for handling messages when receiving messages
+func (sp *Peer) handleMsg(ctx context.Context, msg interface{}) error {
+	switch msg := msg.(type) {
+
+	case *ChequeRequestMsg:
+		return sp.handleChequeRequestMsg(ctx, msg)
+
+	case *EmitChequeMsg:
+		return sp.handleEmitChequeMsg(ctx, msg)
+
+	case *ErrorMsg:
+		return sp.handleErrorMsg(ctx, msg)
+
+	case *ConfirmMsg:
+		return sp.handleConfirmMsg(ctx, msg)
+	}
+
+	return nil
+}
+
+// handleChequeRequestMsg runs when a peer receives a `ChequeRequestMsg`
+// It is thus run by the debitor
+// So the debitor needs to:
+//   * check that it indeed owes to the requestor (if not, ignore message)
+//   * check serial number
+//   * check amount
+//   * if all is ok, issue the cheque
+func (sp *Peer) handleChequeRequestMsg(ctx context.Context, msg interface{}) (err error) {
+	// check we have indeed a negative balance with the peer
+	var req *ChequeRequestMsg
+	var ok bool
+	var peerBalance int64
+
+	if req, ok = msg.(*ChequeRequestMsg); !ok {
+		return fmt.Errorf("Unexpected message type: %v", err)
+	}
+
+	peer := req.Peer
+
+	sp.swap.lock.Lock()
+	defer sp.swap.lock.Unlock() //TODO: Do we really want to block so long?
```

Comment:
Does this mean that attempting to handle another cheque request message while inside this function will be impossible? 

What are the consequences of this lock?

Reply:
Yes - a lock is a semaphore, thus the code between the lock and unlock can not be run in parallel. 

We need to do this because we can not allow that while we are creating the cheque (and thus looking at the balance) another message could be modifying the balance.  It should be looked at as an atomic (or transactional) operation.

It is currently done in a clunky and rudimentary way. It probably needs to be optimized. Feel free to optimize if you can see ways to do so, however keep in mind the concurrency and transactional nature of changes.

Otherwise let's keep this in the back and optimize later.

Reply:
I think at some point we can look into locking data per peer, instead of the whole set of balances, no?

I will make a note of this but for now I think we should keep it as it is.